internal/settings/settings.go

```diff
@@ -0,0 +1,446 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package settings
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+
+	flag "github.com/spf13/pflag"
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/confmap/converter/overwritepropertiesconverter"
+
+	"github.com/signalfx/splunk-otel-collector/internal/configconverter"
+)
+
+const (
+	DefaultUndeclaredFlag = -1
+
+	APIURLEnvVar              = "SPLUNK_API_URL"
+	BallastEnvVarName         = "SPLUNK_BALLAST_SIZE_MIB"
+	ConfigEnvVarName          = "SPLUNK_CONFIG"
+	ConfigServerEnabledEnvVar = "SPLUNK_DEBUG_CONFIG_SERVER"
+	ConfigYamlEnvVarName      = "SPLUNK_CONFIG_YAML"
+	HecLogIngestURLEnvVar     = "SPLUNK_HEC_URL"
+	// nolint:gosec
+	HecTokenEnvVarName    = "SPLUNK_HEC_TOKEN" // this isn't a hardcoded token
+	IngestURLEnvVar       = "SPLUNK_INGEST_URL"
+	MemLimitMiBEnvVarName = "SPLUNK_MEMORY_LIMIT_MIB"
+	MemTotalEnvVarName    = "SPLUNK_MEMORY_TOTAL_MIB"
+	RealmEnvVarName       = "SPLUNK_REALM"
+	// nolint:gosec
+	TokenEnvVarName      = "SPLUNK_ACCESS_TOKEN" // this isn't a hardcoded token
+	TraceIngestURLEnvVar = "SPLUNK_TRACE_URL"
+
+	DefaultGatewayConfig           = "/etc/otel/collector/gateway_config.yaml"
+	DefaultOTLPLinuxConfig         = "/etc/otel/collector/otlp_config_linux.yaml"
+	DefaultLocalGatewayConfig      = "cmd/otelcol/config/collector/gateway_config.yaml"
+	DefaultLocalOTLPLinuxConfig    = "cmd/otelcol/config/collector/otlp_config_linux.yaml"
+	DefaultMemoryBallastPercentage = 33
+	DefaultMemoryLimitPercentage   = 90
+	DefaultMemoryTotalMiB          = 512
+)
+
+type Settings interface {
+	// ResolverURIs returns the collector config provider resolver uris for the collector service
+	ResolverURIs() []string
+	// ConfMapConverters returns the collector config provider resolver confmap.Converters for the collector service
+	ConfMapConverters() []confmap.Converter
+	// ServiceArgs are the sanitized, adjusted args to be used in updating os.Args[1:] for the collector service
+	ServiceArgs() []string
+}
+
+func New(args []string) (Settings, error) {
+	f, err := newFlags(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// immediate exit paths, no further setup required
+	if f.helpFlag || f.versionFlag {
+		return f, nil
+	}
+
+	if err = checkRuntimeParams(f); err != nil {
+		return nil, err
+	}
+
+	if err = setDefaultEnvVars(); err != nil {
+		return nil, err
+	}
+
+	return f, nil
+}
+
+var _ Settings = (*flags)(nil)
+
+type flags struct {
+	configPaths       *stringArrayFlagValue
+	setProperties     *stringArrayFlagValue
+	serviceArgs       []string
+	memBallastSizeMiB int
+	helpFlag          bool
+	versionFlag       bool
+	noConvertConfig   bool
+}
+
+func (f *flags) ResolverURIs() []string {
+	var configPaths []string
+	if configPaths = f.configPaths.value; len(configPaths) == 0 {
+		if configEnvVal := os.Getenv(ConfigEnvVarName); len(configEnvVal) != 0 {
```

Comment:
Long term, might be nice to separate Getenv and Setenv from the settings code since reading/writing them is effectively I/O, and instead pass this information around in some kind of abstraction that could be swappable for testing. Removes the need to deal with global (env) variables in tests.